### PR TITLE
[bridge 12/n] add BridgeActionExecutor

### PR DESCRIPTION
**Pull request by longbowlu**
## Description 

This PR adds BridgeActionExecutor. It takes actions submitted by Orchestrator and does two things:
1. uses AuthorityAggregator to collect signatures
2. submit the signatures on Move.
For simplicity today it only uses 1 gas object, so the Sui tx execution part is handled one by one.

**Retry Logic**
1. If everything goes well (sui tx succeeds), the action will be removed from WAL
2. If sui tx was executed but failed, we give an error log. In this case manual intervention. The only acceptable reason to fail in execution is running out of gas. Otherwise it indicates bugs. Either requires human to check.
4. If sui tx failed in signing stage because of gas object stale, then we refresh the gas object and re-enqueue the action
5. If sui tx failed in signing stage because of other reasons, we gives an error log asking for manual intervention.

## Test Plan 

Unit tests

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
add BridgeActionExecutor

### CODE REVIEW

File: crates/sui-bridge/src/action_executor.rs [hunk 1]
```diff
@@ -0,0 +1,894 @@
...
+pub async fn submit_to_executor(
+    tx: &mysten_metrics::metered_channel::Sender<BridgeActionExecutionWrapper>,
+    action: BridgeAction,
+) -> Result<(), BridgeError> {
+    tx.send(BridgeActionExecutionWrapper(action, 0))
+        .await
+        .map_err(|e| BridgeError::Generic(e.to_string()))
+}
+
+pub fn build_transaction(gas_object_ref: &ObjectRef) -> TransactionData {
```
Comment (longbowlu) [inline crates/sui-bridge/src/action_executor.rs:373]: this is dummy for testing. Will be replaced with actual tx building once the function signature is settled on move.

**Comment by longbowlu** [comment]
cc @Jayfromthe13th

File: crates/sui-bridge/src/action_executor.rs [hunk 2]
```diff
@@ -0,0 +1,894 @@
...
+    ) -> (
+        Vec<tokio::task::JoinHandle<()>>,
+        mysten_metrics::metered_channel::Sender<BridgeActionExecutionWrapper>,
+    );
+}
+
+pub struct BridgeActionExecutor<C> {
+    sui_client: Arc<SuiClient<C>>,
+    bridge_auth_agg: Arc<BridgeAuthorityAggregator>,
+    key: Option<SuiKeyPair>,
```
Comment (lxfind) [inline crates/sui-bridge/src/action_executor.rs:49]: Why Option?

File: crates/sui-bridge/src/action_executor.rs [hunk 3]
```diff
@@ -0,0 +1,894 @@
...
+            info!("Received certified action for execution: {:?}", certificate);
+            let tx_data = build_transaction(&gas_object_ref);
+            let sig = Signature::new_secure(
+                &IntentMessage::new(Intent::sui_transaction(), &tx_data),
+                &sui_key,
+            );
+            let signed_tx = Transaction::from_data(tx_data, vec![sig]);
+            let tx_digest = *signed_tx.digest();
+            info!(?tx_digest, ?gas_object_ref, "Sending transaction to Sui");
+            // TODO: add metrics to detect low balances and so on
```
Comment (lxfind) [inline crates/sui-bridge/src/action_executor.rs:222]: In order to do so you would end up querying the latest object state anyway. So I think doing that as part of the retry loop will simplify some of your logic

File: crates/sui-bridge/src/sui_client.rs [hunk 4]
```diff
@@ -224,6 +255,52 @@ impl SuiClientInner for SuiSdkClient {
...
+        tx: Transaction,
+    ) -> Result<SuiTransactionBlockResponse, BridgeError> {
+        match self.quorum_driver_api().execute_transaction_block(
+            tx,
+            SuiTransactionBlockResponseOptions::new().with_effects(),
+            Some(sui_types::quorum_driver_types::ExecuteTransactionRequestType::WaitForEffectsCert),
+        ).await {
+            Ok(response) => Ok(response),
+            Err(sui_sdk::error::Error::UserInputError(err)) => {
+                if matches!(err, UserInputError::ObjectVersionUnavailableForConsumption{..}) {
```
Comment (lxfind) [inline crates/sui-bridge/src/sui_client.rs:270]: It's a bit hard to reason about whether we have covered all the interesting cases here. I think we need something simpler

File: crates/sui-bridge/src/types.rs [hunk 5]
```diff
@@ -313,6 +316,14 @@ impl Message for BridgeAction {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EthLog {
+    pub block_number: u64,
+    pub tx_hash: H256,
+    pub log_index_in_tx: u16,
+    pub log: Log,
```
Comment (lxfind) [inline crates/sui-bridge/src/types.rs:324]: Do we still need Log here or could we just keep the essentials?

**Review by lxfind — APPROVED** [review]
Approve as overall it's in a good shape to iterate further, but there are a lot of TODOs. I think before we launch anything we would need to:
1. Address all the TODOs
2. Document well and able to reason well about liveness under crash recovery
3. Able to reason about transaction execution flow and gas management really well

File: crates/sui-bridge/src/sui_client.rs (hunk 4 above)
Comment (longbowlu) [reply crates/sui-bridge/src/sui_client.rs:270]: If we follow what you suggested then I think the categorization is not necessary, at least in the beginning

File: crates/sui-bridge/src/action_executor.rs (hunk 2 above)
Comment (longbowlu) [reply crates/sui-bridge/src/action_executor.rs:49]: it's from past iterations, not longer needed, will remove the option

File: crates/sui-bridge/src/types.rs (hunk 5 above)
Comment (longbowlu) [reply crates/sui-bridge/src/types.rs:324]: we can pull the essentials out and flat it. let me add a TODO here to clean it up (saving me some efforts for rebasing)

**Pull request merged**